### PR TITLE
PE-5193: feat(gifting)

### PR DESCRIPTION
**Pull request by thiagocarvalhodev**
- implements the icon for redeem or send gifts
- implement the modal, bloc and method on payment service with the unit tests
- updates the ardrive_http

--- Releases ---
Android release: https://appdistribution.firebase.google.com/testerapps/1:305132849030:android:6cf0cd5ec064fad3ffce07/releases/2qhc3ekjb8bmg

### CODE REVIEW

File: lib/app_shell.dart [hunk 1]
```diff
@@ -179,6 +180,7 @@ class AppShellState extends State<AppShell> {
       setState(() => _showProfileOverlay = !_showProfileOverlay);
 }
 
+// TODO: add the gift icon
```
Comment (matibat) [inline lib/app_shell.dart:183]: Still a TODO?

File: lib/gift/bloc/redeem_gift_bloc.dart [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    required PaymentService paymentService,
+    required ArDriveAuth auth,
+  })  : _auth = auth,
+        _paymentService = paymentService,
+        super(RedeemGiftInitial()) {
+    on<RedeemGiftEvent>((event, emit) async {
+      if (event is RedeemGiftLoad) {
+        await _handleRedeemGiftLoad(emit, event);
+      }
+    });
```
Comment (matibat) [inline lib/gift/bloc/redeem_gift_bloc.dart:24]: No else? If there's no other event we can directly set the callback for the event explicitly.
```suggestion
    on<RedeemGiftLoad>(_handleRedeemGiftLoad);
```

File: lib/gift/bloc/redeem_gift_bloc.dart
Comment (matibat) [inline lib/gift/bloc/redeem_gift_bloc.dart:1]: TODO: cleanup the log statements.

File: lib/gift/bloc/redeem_gift_event.dart [hunk 3]
```diff
@@ -0,0 +1,21 @@
+part of 'redeem_gift_bloc.dart';
+
+sealed class RedeemGiftEvent extends Equatable {
```
Comment (matibat) [inline lib/gift/bloc/redeem_gift_event.dart:3]: I didn't know this of keyword. I'm not 100% sure how does it work but it seems to me that it's behaving as an abstract class from what I can see.

File: lib/gift/reedem_button.dart [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+  Widget build(BuildContext context) {
+    return ArDriveDropdown(
+      anchor: const Aligned(
+        follower: Alignment.topRight,
+        target: Alignment.bottomRight,
+      ),
+      items: [
+        ArDriveDropdownItem(
+          onClick: () {
+            openUrl(url: 'http://gift.ardrive.io/');
```
Comment (matibat) [inline lib/gift/reedem_button.dart:27]: We can put this link in the resources as a constant.

File: lib/l10n/app_en.arb [hunk 5]
```diff
@@ -198,6 +198,10 @@
   "@closeEmphasized": {
     "description": "Close modal. Emphasized with upper case"
   },
+  "codeAlreadyUsed": "Code already used",
+  "@codeAlreadyUsed": {},
```
Comment (matibat) [inline lib/l10n/app_en.arb:202]:
```suggestion
  "@codeAlreadyUsed": {
    "description": "The given code was already used"
  },
```

File: lib/l10n/app_en.arb [hunk 6]
```diff
@@ -198,6 +198,10 @@
...
+  "codeAlreadyUsedDescription": "The code provided has already been redeemed.",
+  "@codeAlreadyUsedDescription": {},
```
Comment (matibat) [inline lib/l10n/app_en.arb:204]:
```suggestion
  "@codeAlreadyUsedDescription": {
    "description": "Explains the code was already been redeemed"
  },
```

File: lib/l10n/app_en.arb [hunk 7]
```diff
@@ -210,6 +214,8 @@
   "@computingSnapshotData": {
     "description": "Computing Snapshot Data"
   },
+  "confirm": "Confirm",
+  "@confirm": {},
```
Comment (matibat) [inline lib/l10n/app_en.arb:218]:
```suggestion
  "@confirm": {
    "description": "The action of confirming"
  },
```

File: lib/l10n/app_en.arb [hunk 8]
```diff
@@ -222,6 +228,8 @@
   "@confirmPassword": {
     "description": "Label for repeating the password again"
   },
+  "confirmTheEmailAddressTheGiftWasSentTo": "Please confirm the email address the gift was sent to and enter the gift code to redeem your gift!",
+  "@confirmTheEmailAddressTheGiftWasSentTo": {},
```
Comment (matibat) [inline lib/l10n/app_en.arb:232]:
```suggestion
  "@confirmTheEmailAddressTheGiftWasSentTo": {
    "description": "Asks for the e-mail to be confirmed in order to redeem the gift"
  },
```

File: lib/l10n/app_en.arb [hunk 9]
```diff
@@ -605,6 +617,8 @@
       }
     }
   },
+  "email": "Email",
+  "@email": {},
```
Comment (matibat) [inline lib/l10n/app_en.arb:621]:
```suggestion
  "@email": {
    "description": "E.g. \"this is my Email\""
  },
```

File: lib/l10n/app_en.arb [hunk 10]
```diff
@@ -985,6 +999,10 @@
   "@getAWallet": {
     "description": "Text for the link that redirects the user to the 'Claim a token' page of ArWeave"
   },
+  "giftCode": "Gift Code\n",
+  "@giftCode": {},
```
Comment (matibat) [inline lib/l10n/app_en.arb:1003]: I would say the newline should be applied with code instead.
```suggestion
  "giftCode": "Gift Code",
  "@giftCode": {
    "description": "The code for the gift"
  },
```

File: lib/l10n/app_en.arb [hunk 11]
```diff
@@ -985,6 +999,10 @@
...
+  "giftRedeemed": "Gift Redeemed!",
+  "@giftRedeemed": {},
```
Comment (matibat) [inline lib/l10n/app_en.arb:1005]:
```suggestion
  "@giftRedeemed": {
    "description": "Indicates the gift was sucessfully redeemed"
  },
```

File: lib/l10n/app_en.arb [hunk 12]
```diff
@@ -1077,6 +1095,8 @@
   "@insufficientFundsForUploadFolders": {
     "description": "Show that needs funds to upload folders"
   },
+  "invalidCode": "Invalid code",
+  "@invalidCode": {},
```
Comment (matibat) [inline lib/l10n/app_en.arb:1099]:
```suggestion
  "@invalidCode": {
    "description": "Indicates the given code is invalid"
  },
```

File: lib/l10n/app_en.arb [hunk 13]
```diff
@@ -1508,6 +1532,16 @@
   "@recreatingFolderEmphasized": {
     "description": "Folder is being recreated. Emphasized with upper case"
   },
+  "redeemGift": "Redeem gift",
+  "@redeemGift": {},
```
Comment (matibat) [inline lib/l10n/app_en.arb:1536]:
```suggestion
  "@redeemGift": {
    "description": "The action of redeeming a gift"
  },
```

File: lib/l10n/app_en.arb [hunk 14]
```diff
@@ -1508,6 +1532,16 @@
...
+  "redeemYourGift": "Redeem Your Gift",
+  "@redeemYourGift": {},
```
Comment (matibat) [inline lib/l10n/app_en.arb:1538]:
```suggestion
  "@redeemYourGift": {
    "description": "The action of redeeming your gift"
  },
```

File: lib/l10n/app_en.arb [hunk 15]
```diff
@@ -1508,6 +1532,16 @@
...
+  "redemptionErrorDescription": "Redemption was unsuccessful",
+  "@redemptionErrorDescription": {},
```
Comment (matibat) [inline lib/l10n/app_en.arb:1540]:
```suggestion
  "@redemptionErrorDescription": {
    "description": "Indicates there was a problem while redempting"
  },
```

File: lib/l10n/app_en.arb [hunk 16]
```diff
@@ -1508,6 +1532,16 @@
...
+  "redemptionErrorTitle": "No soup for you!",
+  "@redemptionErrorTitle": {},
```
Comment (matibat) [inline lib/l10n/app_en.arb:1542]:
```suggestion
  "@redemptionErrorTitle": {
    "description": "You get no soup! >:)"
  },
```

File: lib/l10n/app_en.arb [hunk 17]
```diff
@@ -1508,6 +1532,16 @@
...
+  "redemptionSuccessDescription": "Your credits will be added to your account soon. You can close this window.",
+  "@redemptionSuccessDescription": {},
```
Comment (matibat) [inline lib/l10n/app_en.arb:1544]:
```suggestion
  "@redemptionSuccessDescription": {
    "description": "Indicates the redeemed gift is gonna be available for use soon"
  },
```

File: lib/l10n/app_en.arb [hunk 18]
```diff
@@ -1584,6 +1618,8 @@
   "@selectWalletEmphasized": {
     "description": "Users can log in with a crypto wallet, so this prompts them to choose one for their login. Emphasized with upper case"
   },
+  "sendGift": "Send gift",
+  "@sendGift": {},
```
Comment (matibat) [inline lib/l10n/app_en.arb:1622]:
```suggestion
  "@sendGift": {
    "description": "The action of sending a gift"
  },
```

File: lib/l10n/app_en.arb [hunk 19]
```diff
@@ -1744,6 +1792,8 @@
   "@termsAndConditions": {
     "description": "terms and conditions"
   },
+  "theGiftCodeProvidedIsInvalid": "The gift code provided is invalid.",
+  "@theGiftCodeProvidedIsInvalid": {},
```
Comment (matibat) [inline lib/l10n/app_en.arb:1796]:
```suggestion
  "@theGiftCodeProvidedIsInvalid": {
    "description": "Indicates the given code is invalid"
  },
```

File: lib/turbo/services/payment_service.dart [hunk 20]
```diff
@@ -128,6 +129,40 @@ class PaymentService {
...
+      if (result.statusCode != 200) {
+        throw Exception(
+            'Gift redeem failed with status code ${result.statusCode}');
+      }
+
+      final newBalance = result.data['userBalance'] as String;
+
+      return int.parse(newBalance);
+    } on ArDriveHTTPException catch (e) {
+      if (e.data == 'Gift has already been redeemed!') {
```
Comment (matibat) [inline lib/turbo/services/payment_service.dart:156]: We should use a more reliable way of determining what the error is. If this is the only way then we should create a tech debt for later.

File: packages/ardrive_utils/lib/src/validations.dart
Comment (matibat) [inline packages/ardrive_utils/lib/src/validations.dart:1]: I'm pretty sure we had this method somewhere else, we should use that. If this is the same we previously had then that's fine. And, we should have unit tests for it.

File: lib/app_shell.dart (hunk 1 above)
Comment (thiagocarvalhodev) [reply lib/app_shell.dart:183]: No, it's done.

File: lib/gift/bloc/redeem_gift_event.dart (hunk 3 above)
Comment (thiagocarvalhodev) [reply lib/gift/bloc/redeem_gift_event.dart:3]: https://dart.dev/language/class-modifiers

File: lib/gift/reedem_button.dart (hunk 4 above)
Comment (thiagocarvalhodev) [reply lib/gift/reedem_button.dart:27]: 30c8e484215d21b03a876c231ccfcb0c4ee2b7eb

File: lib/turbo/services/payment_service.dart (hunk 20 above)
Comment (thiagocarvalhodev) [reply lib/turbo/services/payment_service.dart:156]: Yes, I completely agree. I'll create the tech debt

File: lib/gift/bloc/redeem_gift_bloc.dart
Comment (thiagocarvalhodev) [reply lib/gift/bloc/redeem_gift_bloc.dart:1]: I think it's ok having the log statements there as they are debug level.

**Review by matibat — DISMISSED** [review]
LGTM

**Review by matibat — APPROVED** [review]

**Pull request merged**